onnxruntime/contrib_ops/cpu/transformers/logits_processor.cc

```diff
@@ -0,0 +1,192 @@
+      T score = beam_token_scores[word_id];
+
+      // If score < 0, then repetition penalty > 1.0 has to multiplied to reduce the previous token probability,
+      // This assumes that scores are either positive (like ctrl) or negative (like GPT-2), but not a mixture.
+      beam_token_scores[word_id] = (score < 0 ? score * penalty_ : score / penalty_);
+    }
+  }
+
+#ifdef DEBUG_BEAM_SEARCH
+  DumpScores("RepetitionPenaltyLogitsProcessor", next_token_scores.scores);
+#endif
+}
+
+template <typename T>
+NoRepeatNGramLogitsProcessor<T>::NoRepeatNGramLogitsProcessor(int ngram_size) : ngram_size_(ngram_size) {
+}
+
+template <typename T>
+void NoRepeatNGramLogitsProcessor<T>::Process(const ISequences* sequences,
+                                              NextTokenScores<T>& next_token_scores) {
+  if (ngram_size_ == 0 || ngram_size_ > sequences->GetSequenceLength()) {
+    return;
+  }
+
+  const gsl::index prefix_length = static_cast<gsl::index>(ngram_size_ - 1);
+  int batch_beam_size = next_token_scores.batch_beam_size;
+
+  for (int i = 0; i < batch_beam_size; i++) {
+    gsl::span<T> beam_token_scores = next_token_scores.GetScores(i);
+    gsl::span<const int64_t> sequence = sequences->GetSequence(i);
+
+    gsl::span<const int64_t> prefix = sequence.subspan(sequence.length() - prefix_length);
+    ORT_ENFORCE(prefix.length() == prefix_length);
+
+    std::unordered_set<int64_t> blocked_word_ids;
+    for (int j = 0; j <= static_cast<int>(sequence.length()) - ngram_size_; j++) {
+      // Here we use naive algorithm for matching. The complexity is O(batch_beam_size * ngram_size * sequence_length)
+      // TODO: build N-Gram index (hash table with prefix of length NGram - 1 as key, and list of last word of NGram as value) for fast matching.
+      if (ngram_size_ == 1 || prefix == sequence.subspan(j, prefix_length)) {
+        blocked_word_ids.insert(sequence[j + prefix_length]);
+      }
+    }
+
+    for (const int64_t word_id : blocked_word_ids) {
+      beam_token_scores[word_id] = std::numeric_limits<T>::lowest();
+    }
+  }
+
+#ifdef DEBUG_BEAM_SEARCH
+  DumpScores("NoRepeatNGramLogitsProcessor", next_token_scores.scores);
+#endif
+}
+
+template <typename T>
+VocabMaskLogitsProcessor<T>::VocabMaskLogitsProcessor(const gsl::span<const int32_t>& vocab_mask) : vocab_mask_(vocab_mask) {
+}
+
+template <typename T>
+void VocabMaskLogitsProcessor<T>::Process(const ISequences* /*sequences*/,
+                                          NextTokenScores<T>& next_token_scores) {
+  assert(!vocab_mask_.empty());
+
+  // Process vocabulary mask and set tokens with mask value 0 to -inf.
+  T* p = next_token_scores.scores.data();
+  // next_token_scores shape (batch_size * num_beams, vocab_size)
+  // vocab_mask shape (vocab_size). TODO: support shape (batch_size, vocab_size)
+  for (int i = 0; i < next_token_scores.batch_beam_size; i++) {
+    for (int j = 0; j < next_token_scores.vocab_size; j++, p++) {
```

Comment:
We are applying same vocab mask for all the batches.
This means all the sequences have same vocab mask. One use case I think of vocab mask is where it is prepared by tokenzier to mark the relevant tokens in next inference(the first one). Each batch can have different seq_len and different vocab mask.
Is this done to address a diff case?

Reply:
That's the reason I added a comment above the code "TODO: support shape (batch_size, vocab_size)".

There are two scenarios: one case is to used as bad word list as in huggingface transformers, that list is global. Another case is per batch mask. That can be done in another PR.

Reply:
Internal TODOs are good for devs. My question is more on the restriction this brings in. Are you documenting this? What can we provide as reference to users?

Reply:
Most TODOs shall be resolved before next release 1.11. For reference, the definition of the operator has shape information for each input.